### ISSUE TITLE
vue/v9: css modules cannot co-exist with global styles (could in v8)

### ISSUE THREAD

**Issue by timkelty** [neutrinojs/neutrino#1243]
In v8, it was possible to have vue use css modules, but have co-existing global (non-module) styles.

With the same config in v9, this doesn't work:
```
Module parse failed: Unexpected token (19:0)
You may need an appropriate loader to handle this file type.
```

Working v8 example: https://github.com/timkelty/vue-scss/tree/v8-working
Broken v8 example: https://github.com/timkelty/vue-scss/tree/v9-broken

You can still achieve this in v9, but have to do thing differently:

1. set `modulesTest` equal to `test` (https://github.com/neutrinojs/neutrino/pull/1242 should help with that)
2. use `:global` in your css where you want global (non-module) styles.

-----

If using modules everywhere and requiring `:global` is the only way to do it in v9, that's fine, but it should probably be noted somewhere, as you could do it in v8. A the very least I think we should merge in https://github.com/neutrinojs/neutrino/pull/1242 so you don't have to set both `test` and `modulesTest`

**Comment by edmorley** [comment]
I'm presuming this is as a result of the changes in #820? Specifically this?
https://github.com/neutrinojs/neutrino/pull/820/files#diff-8e01b2cd41edef3f12c383a2371bc95bR7

I agree #1242 sounds like the right thing to do with the current implementation. However beyond that I'm not very familiar with how Vue configures styles and am pretty overloaded with things to work on at the moment, so probably won't have a chance to look at this unfortunately.

**Comment by timkelty** [comment]
@edmorley Yep, I think that its the culprit.

I'll take a look today.

**Comment by timkelty** [comment]
I did a little more digging and included webpack configs from vue-cli, which works as expected:

- output from vue-cli: https://github.com/timkelty/vue-scss/blob/master/vue-cli.js
- output from v9/neutrino: https://github.com/timkelty/vue-scss/blob/master/v9-broken.config

To summarize, the goal is to:

- by default, parse css as global (modules: false)
- in a .vue component, parse `<style module>` as css modules
- in a .js, parse `import 'styles.css'` as global and `import 'styles.module.css'` as css module

**Comment by timkelty** [comment]
https://github.com/vuejs/vue-cli/blob/b559005ce64dc801c280e38194f28dd583292095/packages/%40vue/cli-service/lib/config/css.js#L72

Here is what vue-cli is doing (it uses webpack-chain under the hood)

**Comment by timkelty** [comment]
So I have a functioning fix for this, but it seems really ugly: https://github.com/timkelty/neutrino-dev/commit/f19cb7046ed912a5c3f5adcd2da26dcd33920e59

I think theres just some webpack-chain-fu I can't quite figure out:

https://github.com/neutrinojs/webpack-chain/issues/119
https://github.com/neutrinojs/webpack-chain/issues/129

**Comment by edmorley** [comment]
Yeah using `oneOfs` sounds like a good move :-)

**Comment by timkelty** [comment]
@edmorley wont take much to refactor style-loader to use oneOfs, I can include that…

However, we're going to run into [this](https://github.com/neutrinojs/webpack-chain/issues/119), since the added rules for vue will need to come first. Any ideas there?

**Comment by edmorley** [comment]
@timkelty I'll release a new version of webpack-chain that includes the `oneOf` `.before()` support soon (just waiting on a few other webpack-chain PRs first). In the meantime would you have a chance to open a separate PR for switching `@neutrinojs/style-loader` to using `oneOf`s? We can then get that reviewed/merged and rebase #1276 on top of it.

**Comment by timkelty** [comment]
@edmorley will do!
